Chatto/Tests/ObservableAsyncStreamTests.swift

```diff
@@ -0,0 +1,73 @@
+//
+// The MIT License (MIT)
+//
+// Copyright (c) 2015-present Badoo Trading Limited.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import Foundation
+import XCTest
+import Chatto
+
+@available(iOS 13, *)
+final class ObservableAsyncStreamTests: XCTestCase {
+
+    func test_GivenStreamFromObservable_WhenObservableUpdates_ThenValueDeliveredToStream() async throws {
+        let observable = Observable(0)
+        let stream = observable.asyncStream()
+        async let firstReceivedValue: Int? = stream.first { _ in true }
+        observable.value = 1
+        let result = await firstReceivedValue
+        XCTAssertEqual(result, 1)
+    }
+
+    func test_WhenObserverDeallocated_ThenStreamCompletes() async throws {
+        var observable: Observable<Void>? = .init(())
+        let stream = observable!.asyncStream()
+        let iterateAllStreamValuesTask = Task {
+            for await _ in stream {}
+        }
+        observable = nil
+
+        try await withTimeout {
+            await withTaskCancellationHandler {
+                await iterateAllStreamValuesTask.value
+            } onCancel: {
+                iterateAllStreamValuesTask.cancel()
+            }
+        }
```

Comment:
It feels a bit counter-intuitive that we're using `withTaskCancellationHandler` which handles cancellation, to trigger a cancellation to test it out.

May I suggest something like following to make it more clear that we are triggering cancel and testing that stream finishes?
```swift
var (finishStream, finishContinuation) = AsyncStream.makeStream(of: Void.self)
let waitForFinish: (_ stream: AsyncStream<Void>) async -> Void = { for await _ in $0 { } }

let iterateAllValuesTask = Task {
    await waitForFinish(stream)
    finishContinuation.finish()
}

finishContinuation.onTermination = {
    XCTAssertEqual($0, .finished, "Test finished before observer finishing the underlying AsyncStream")
}

iterateAllValuesTask.cancel()
await waitForFinish(finishStream)
XCTAssertEqualNil(await finishStream.first { _ in true }) // to make sure stream is finished and now is returning nil
```

We can also omit the last line as sdk guarantees `onTermination` is called in this situation before finishing an iteration so we have a chance for clean-up.

Reply:
Oh, the `withTaskCancellationHandler` isn't part of the logic I'm testing there.

The test is checking if stream will finish when Observable is deallocated.

I've added the cancellation handler to cancel my unstructured task, when the task running test case itself being cancelled.

Reply:
But to be frank, I don't think it's needed, since I'm going to cancel it with timeout anyway. Do you think I should remove it, since it brings the confusion

Reply:
Yeah I figured that part, but my 2 cents, it's still better to see we call `cancel` on the sut Task explicitly, rather than calling it with a time-out.